soc: renesas: ra: configure option settings memory

iandmorris opened the pull request:
An area of flash memory on the RA4M1 MCU is used to store information used to configure the device following a reset. This patch instructs the linker to reserve this memory area and provides kconfig options that are used to populate it (at build time) with the desired device configuration.

NOTES:

The only board that currently uses the RA4M1 soc port is the Ardunio UNO R4 Minima. This board uses a bootloader to load the application. As this bootloader is located in the lower part of the RA4M1 flash memory it handles the reservation (and configuration) of the option setting memory area. In this case (and others where the application does not reside in the area that contains the option setting memory) this functionality can be disabled via the OPTION_SETTING_MEMORY kconfig option, however leaving it enabled does not cause any issues other than wasting 256 bytes of flash memory.

This option MUST be enabled if the application resides in the area that contains the option setting memory (which is usually the case if a bootloader is not used) - otherwise unintended behavior can occur due to misconfiguration of the device.

iandmorris committed:
soc: renesas: ra: configure option settings memory

An area of flash memory on the RA4M1 MCU is used to store information
used to configure the device following a reset. This patch instructs
the linker to reserve this memory area and provides kconfig options
that are used to populate it (at build time) with the desired device
configuration.

Signed-off-by: Ian Morris <ian.d.morris@outlook.com>

thedjnK commented on soc/renesas/ra/ra4m1/Kconfig.soc:
```diff
@@ -18,3 +18,27 @@ config SOC_SERIES
 config SOC
 	default "r7fa4m1ab3cfm" if SOC_R7FA4M1AB3CFM
+
+config OPTION_SETTING_MEMORY
+	bool "Option Setting Memory"
+	default y
+
+config OPTION_SETTING_LVDAS_DISABLE
+	bool "Disable voltage monitor 0 after reset"
+	default y
+	depends on OPTION_SETTING_MEMORY
+
+config OPTION_SETTING_HOCO_DISABLE
+	bool "Disable HOCO oscillator after reset"
+	default n
+	depends on OPTION_SETTING_MEMORY
+
+config OPTION_SETTING_HOCO_FREQ
+	int "HOCO oscillator frequency"
+	default 48000000
+	depends on OPTION_SETTING_MEMORY
+
+config OPTION_SETTING_VDSEL
+	int "Voltage Detection 0 Level Select"
+	default 1900
+	depends on OPTION_SETTING_MEMORY
```
Move to `Kconfig`

soburi commented on soc/renesas/ra/ra4m1/Kconfig.soc:
```diff
@@ -18,3 +18,27 @@ config SOC_SERIES
+	default y
+	depends on OPTION_SETTING_MEMORY
+
+config OPTION_SETTING_HOCO_DISABLE
```
Please refer to the devicetree definitions.

soburi replied:
https://github.com/zephyrproject-rtos/zephyr/blob/f643f3a6cf6bafd6338c090f4413e34eb3ed806c/dts/arm/renesas/ra/ra-cm4-common.dtsi#L39-L44

Please use this definition for determining HOCO frequency.

thedjnK commented on soc/renesas/ra/ra4m1/Kconfig:
```diff
@@ -9,3 +9,27 @@ config SOC_SERIES_RA4M1
+
+config OPTION_SETTING_HOCO_DISABLE
+	bool "Disable HOCO oscillator after reset"
+	default n
```
`default n` not needed

thedjnK commented on soc/renesas/ra/ra4m1/Kconfig:
```diff
@@ -9,3 +9,27 @@ config SOC_SERIES_RA4M1
+	default n
+	depends on OPTION_SETTING_MEMORY
+
+config OPTION_SETTING_HOCO_FREQ
+	int "HOCO oscillator frequency"
+	default 48000000
+	depends on OPTION_SETTING_MEMORY
+
+config OPTION_SETTING_VDSEL
+	int "Voltage Detection 0 Level Select"
+	default 1900
+	depends on OPTION_SETTING_MEMORY
```
these should be dts entries

thedjnK commented on soc/renesas/ra/ra4m1/Kconfig:
```diff
@@ -9,3 +9,27 @@ config SOC_SERIES_RA4M1
 	select TIMER_READS_ITS_FREQUENCY_AT_RUNTIME
 	select XIP
+
+config OPTION_SETTING_MEMORY
```
settings here should have `SOC_` prefix

soburi commented on soc/renesas/ra/ra4m1/soc.c:
```diff
@@ -0,0 +1,148 @@
+#endif
+
+struct ofs0_reg {
+	uint32_t rsvd1: 1;
```
Please capitalize it like others

soburi commented on soc/renesas/ra/ra4m1/soc.c:
```diff
@@ -0,0 +1,148 @@
+#ifdef CONFIG_SOC_OPTION_SETTING_MEMORY
+const struct opt_set_mem ops __attribute__((section(".opt_set_mem"))) = {
+	.ofs0 = {
+		/*
```
I couldn't estimate what the design intention from the "Set all fields to 1” description.
Since WDTSTPCTL is 1, the watchdog will not start, so it would be better to state that.

soburi commented on soc/renesas/ra/ra4m1/Kconfig:
```diff
@@ -9,3 +9,26 @@ config SOC_SERIES_RA4M1
+	bool "Option Setting Memory"
+	default y
+
+config SOC_OPTION_SETTING_LVDAS_DISABLE
```
The voltage detecting and enabling HOCO on reset setting seems good to be managed by hwinfo driver.
If you are having difficulty creating a suitable devicetree at the moment,
I think it's okay to set a fixed value at this point.

soburi commented:
@iandmorris 

I have created PR #71174 based on these patches.
I notice that *The option setting is a common function in the RA series.*
So, I think defining in a standard file in the RA series is good.

soburi commented on soc/renesas/ra/ra4m1/Kconfig:
```diff
@@ -9,3 +9,26 @@ config SOC_SERIES_RA4M1
+	bool "Disable HOCO oscillator after reset"
+	depends on SOC_OPTION_SETTING_MEMORY
+
+config SOC_OPTION_SETTING_HOCO_FREQ
+	int "HOCO oscillator frequency"
+	default $(dt_node_int_prop_int,/clocks/hoco,clock-frequency)
+	depends on SOC_OPTION_SETTING_MEMORY
```
We can use the devicetree definition to determine hoco frequency by the following code.

```
DT_PROP(DT_PATH(clocks, hoco), clock_frequency)
```

iandmorris replied:
hi @soburi, i'm new to zephyr so i could be wrong about this but is seems like I can't use:

```DT_PROP(DT_PATH(clocks, hoco), clock_frequency)```

inside a Kconfig file to access a devicetree property, however i can use this within a 'C' file to access a devicetree property.

when i check other Kconfig files that refer to the devicetree they all seem to use:

```dt_node_xxx_yyy```

see soc/nxp/mcx/mcxnx4x/Kconfig for an example

let me know what you think, I probably missed something obvious :)

soburi replied:
I think this part ...

https://github.com/zephyrproject-rtos/zephyr/pull/71017/files#diff-7cd037a8cbd8c3e02942d0cbda9cf90aa08fbab8ab55dcf57b46e12f660662beR8-R18

can rewrite as follows.
We can refer the hoco clock frequency directly in code, so we no need to define hoco freq in Kconfig.

```
#define HOCO_FREQ DT_PROP(DT_PATH(clocks, hoco), clock_frequency)

#if HOCO_FREQ == 24000000
#define OFS1_HOCO_FREQ		0
#elif HOCO_FREQ == 32000000
#define OFS1_HOCO_FREQ		2
#elif HOCO_FREQ == 48000000
#define OFS1_HOCO_FREQ		4
#elif HOCO_FREQ == 64000000
#define OFS1_HOCO_FREQ		5
#else
#error "Unsupported HOCO frequency"
#endif
```

soburi commented on soc/renesas/ra/ra4m1/Kconfig:
```diff
@@ -9,3 +9,26 @@ config SOC_SERIES_RA4M1
+	default y
+	depends on SOC_OPTION_SETTING_MEMORY
+
+config SOC_OPTION_SETTING_HOCO_DISABLE
```
Whether the hoco is valid or invalid, I don't think there is any problem with judging it by the following code.

````
DT_NODE_HAS_STATUS(DT_PATH(clocks, hoco), okay)
````

If you are considering a case where the not enabled the hoco at reset but is enabled during startup processing, add a property to `dts/bindings/clock/renesas,ra-clock-generation-circuit.yaml`. I think it's also good. (This might be overthinking.)

soburi replied:
This is same as `SOC_OPTION_SETTING_HOCO_FREQ`.

https://github.com/zephyrproject-rtos/zephyr/pull/71017/files#diff-7cd037a8cbd8c3e02942d0cbda9cf90aa08fbab8ab55dcf57b46e12f660662beR122

This line can rewrite 

```
		.HOCOEN = !DT_NODE_HAS_STATUS(DT_PATH(clocks, hoco), okay),
```

And we can remove this Kconfig option.

soburi replied:
https://github.com/zephyrproject-rtos/zephyr/pull/71017/files#diff-7cd037a8cbd8c3e02942d0cbda9cf90aa08fbab8ab55dcf57b46e12f660662beR10-R20

It seems to be a good. We can remove the config.

```
config SOC_OPTION_SETTING_HOCO_DISABLE
```

soburi commented on soc/renesas/ra/ra4m1/Kconfig:
```diff
@@ -9,3 +9,26 @@ config SOC_SERIES_RA4M1
+	default $(dt_node_int_prop_int,/clocks/hoco,clock-frequency)
+	depends on SOC_OPTION_SETTING_MEMORY
+
+config SOC_OPTION_SETTING_VDSEL
```
I think it's better to create `dts/bindings/hwinfo/renesas,ra-lvd.yaml` and define it in the devicetree node.

soburi replied:
I don't think this setting needs to be done in this PR.
Since the hwinfo driver is used to obtain the startup factor after a reset, I think it would be a good idea to create a devicetree definition related to this and use that.
(If you are new to Zephyr development, creating a hwinfo driver brings good experience.)

iandmorris replied:
@soburi, Agreed, I would like to implement this setting via the device tree but do it in another PR. Shall I leave the VDSEL as it is for this PR or maybe remove the option?

soburi replied:
Don't leave options that don't work, as they confuse users.

soburi commented on soc/renesas/ra/ra4m1/soc.c:
```diff
@@ -0,0 +1,148 @@
+/*
```
The OFS is used widely in the RA series.
It should place a `common` directory.

iandmorris replied:
@soburi, while it's true that the OFS is used in a lot of RA microcontrollers it seems the contents of the registers, and their memory location is different in pretty much every series. A quick check of the device datasheets shows that RA6 series has different OSF1 register contents than RA4 series, as does the RA0 series. Therefore, I think it is better to leave this where it is...

soburi replied:
If so, let's make only Kconfig common.

iandmorris commented:
@soburi, i think i have addressed all the comments. let me know if i missed something or if you have anything further.

soburi reviewed (approved):
I would like to make this OK for now.

Renesas has officially started supporting it, so there may be things to consider in the future.

https://github.com/zephyrproject-rtos/zephyr/pull/72150

iandmorris commented:
Hi @thedjnK, I think I have addressed your comments. When you get a minute could you take a look and approve (or let me know if further changes are required)?
Thanks
Ian

thedjnK commented on soc/renesas/ra/ra4m1/Kconfig:
```diff
@@ -9,3 +9,7 @@ config SOC_SERIES_RA4M1
 	select TIMER_READS_ITS_FREQUENCY_AT_RUNTIME
 	select XIP
+
+config SOC_OPTION_SETTING_MEMORY
+	bool "Option Setting Memory"
+	default y
```
Shouldn't this have a description? Never used these parts but "option setting memory" to me has no meaning, I have no idea what it is or what disabling it would do. Non-blocking comment and can be updated in a future PR.
